File: ios/Plugin/CapacitorCrispPlugin.swift
import Foundation
import Capacitor
import Crisp

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorCrispPlugin)
public class CapacitorCrispPlugin: CAPPlugin {
      @objc override public func load() {
        // Called when the plugin is first constructed in the bridge
        print("CapacitorCrispPlugin Initialized")
    }

    @objc func configure(_ call: CAPPluginCall) {
        let websiteID = call.getString("websiteID") ?? ""
        print("Crisp Configure " + websiteID)
        CrispSDK.configure(websiteID: websiteID)
        call.resolve()
    }

    @objc func openMessenger(_ call: CAPPluginCall) {
        print("Open Crisp Widget")
        DispatchQueue.main.async {
            self.bridge?.viewController?.present(ChatViewController(), animated: true, completion: nil)
        }
        call.resolve()
    }

    @objc func setTokenID(_ call: CAPPluginCall) {
        let tokenID = call.getString("tokenID") ?? ""
        CrispSDK.setTokenID(tokenID: tokenID)
        call.resolve()
    }

    @objc func setUser(_ call: CAPPluginCall) {
        let nickname = call.getString("nickname") ?? nil
        let phone = call.getString("phone") ?? nil
        let email = call.getString("email") ?? nil
        let avatar = call.getString("avatar") ?? nil
        if nickname != nil {
            CrispSDK.user.nickname = nickname;
        }
        if phone != nil {
            CrispSDK.user.phone = phone;
        }
        if email != nil {
            CrispSDK.user.email = email;
        }
        if avatar != nil {
            CrispSDK.user.avatar = URL(string: avatar ?? "");
        }
        call.resolve()
    }

    @objc func pushEvent(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? ""
        let color = call.getString("color") ?? ""
        switch color {
            case "red":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.red))
            case "orange":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.orange))
            case "yellow":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.yellow))
            case "green":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.green))
            case "blue":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.blue))
            case "purple":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.purple))
            case "pink":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.pink))
            case "brown":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.brown))
            case "grey":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.grey))
            case "black":
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.black))
            default:
                CrispSDK.session.pushEvent(SessionEvent(name: name, color: SessionEventColor.blue))
        }
        call.resolve()
    }

    @objc func setCompany(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? nil
        let url = URL(string: call.getString("url") ?? "")
        let description = call.getString("description") ?? nil
        let employment = call.getArray("employment", String.self) ?? ["", ""]
        let geolocation = call.getArray("geolocation", String.self) ?? ["", ""]
        CrispSDK.user.company = Company(name: name, url: url, companyDescription: description, 
        employment: Employment(title: employment[0], role: employment[1]),
        geolocation:  Geolocation(city: geolocation[0], country: geolocation[1]))
        call.resolve()
    }

    @objc func setString(_ call: CAPPluginCall) {
        let key = call.getString("key") ?? ""
        let value = call.getString("value") ?? ""
        CrispSDK.session.setString(value, forKey: key)
        call.resolve()
    }

    @objc func setInt(_ call: CAPPluginCall) {
        let key = call.getString("key") ?? ""
        let value = call.getInt("value") ?? 0
        CrispSDK.session.setInt(value, forKey: key)
        call.resolve()
    }

    @objc func setSegment(_ call: CAPPluginCall) {
        let segment = call.getString("segment") ?? ""
        CrispSDK.session.segment = segment
        call.resolve()
    }

    @objc func reset(_ call: CAPPluginCall) {
        CrispSDK.session.reset()
        call.resolve()
    }
}

File: android/src/main/java/ee/forgr/plugin/crisp/CapacitorCrispPlugin.java
package ee.forgr.plugin.crisp;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Build;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import im.crisp.client.external.ChatActivity;
import im.crisp.client.external.Crisp;
import im.crisp.client.external.EventsCallback;
import im.crisp.client.external.data.Company;
import im.crisp.client.external.data.Employment;
import im.crisp.client.external.data.Geolocation;
import im.crisp.client.external.data.SessionEvent;
import im.crisp.client.external.data.message.Message;
import im.crisp.client.external.notification.CrispNotificationClient;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Locale;
import java.util.Map;
import org.json.JSONException;

@CapacitorPlugin(name = "CapacitorCrisp")
public class CapacitorCrispPlugin extends Plugin {

    private final String pluginVersion = "8.1.2";

    protected static final int OPEN_MESSENGER_CODE = 12345; // Unique request code

    private Locale configuredLocale;

    private final EventsCallback eventsCallback = new EventsCallback() {
        @Override
        public void onSessionLoaded(String sessionId) {
            JSObject ret = new JSObject();
            ret.put("sessionId", sessionId);
            notifyCrispEvent("sessionLoaded", ret);
        }

        @Override
        public void onChatOpened() {
            notifyCrispEvent("chatOpened", new JSObject());
        }

        @Override
        public void onChatClosed() {
            notifyCrispEvent("chatClosed", new JSObject());
        }

        @Override
        public void onMessageSent(Message message) {
            notifyCrispEvent("messageSent", getMessageEvent(message));
        }

        @Override
        public void onMessageReceived(Message message) {
            notifyCrispEvent("messageReceived", getMessageEvent(message));
        }
    };

    @Override
    public void load() {
        super.load();
        Crisp.addCallback(this.eventsCallback);
    }

    @Override
    protected void handleOnDestroy() {
        Crisp.removeCallback(this.eventsCallback);
        super.handleOnDestroy();
    }

    private JSObject getMessageEvent(Message message) {
        JSObject ret = new JSObject();
        if (message != null) {
            ret.put("isMe", message.isMe());
        }
        return ret;
    }

    private JSObject getPushMessageEvent() {
        JSObject ret = new JSObject();
        ret.put("isMe", false);
        ret.put("fromPushNotification", true);
        return ret;
    }

    private void notifyCrispEvent(String eventName, JSObject data) {
        if (this.getActivity() != null) {
            this.getActivity().runOnUiThread(() -> notifyListeners(eventName, data));
            return;
        }
        notifyListeners(eventName, data);
    }

    private Context getCrispContext() {
        Context activity = this.getActivity();
        if (activity != null) {
            return activity;
        }
        return this.getContext();
    }

    private Context applyConfiguredLocale(Context context) {
        if (this.configuredLocale == null) {
            return context;
        }
        Locale.setDefault(this.configuredLocale);
        Configuration configuration = new Configuration(context.getResources().getConfiguration());
        configuration.setLocale(this.configuredLocale);
        configuration.setLayoutDirection(this.configuredLocale);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return context.createConfigurationContext(configuration);
        }
        context.getResources().updateConfiguration(configuration, context.getResources().getDisplayMetrics());
        return context;
    }

    private void setLocaleFromTag(String localeTag) {
        if (localeTag == null || localeTag.isEmpty()) {
            this.configuredLocale = null;
            return;
        }
        Locale locale = Locale.forLanguageTag(localeTag);
        if (locale.getLanguage().isEmpty()) {
            locale = new Locale(localeTag);
        }
        if (locale.getLanguage().isEmpty()) {
            return;
        }
        this.configuredLocale = locale;
    }

    @PluginMethod
    public void configure(PluginCall call) {
        Context crispContext = this.getCrispContext();
        String websiteID = call.getString("websiteID");
        String tokenID = call.getString("tokenID");
        String localeTag = call.getString("locale");
        if (websiteID == null || websiteID.isEmpty()) {
            call.reject("websiteID is required");
            return;
        }
        this.setLocaleFromTag(localeTag);
        crispContext = this.applyConfiguredLocale(crispContext);
        if (tokenID != null && !tokenID.isEmpty()) {
            Crisp.configure(crispContext, websiteID, tokenID);
        } else {
            Crisp.configure(crispContext, websiteID);
        }
        Crisp.enableNotifications(crispContext, true);
        call.resolve();
    }

    @PluginMethod
    public void setTokenID(PluginCall call) {
        Context crispContext = this.applyConfiguredLocale(this.getCrispContext());
        String tokenID = call.getString("tokenID");
        if (tokenID == null || tokenID.isEmpty()) {
            call.reject("tokenID is required");
            return;
        }
        Crisp.setTokenID(crispContext, tokenID);
        call.resolve();
    }

    @PluginMethod
    public void openMessenger(PluginCall call) {
        Context crispContext = this.applyConfiguredLocale(this.getCrispContext());
        Intent crispIntent = new Intent(crispContext, ChatActivity.class);
        if (this.getActivity() == null) {
            crispIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        }
        crispContext.startActivity(crispIntent);
        call.resolve();
    }

    @PluginMethod
    public void setUser(PluginCall call) {
        if (call.hasOption("email")) {
            String email = call.getString("email");
            Crisp.setUserEmail(email);
        }
        if (call.hasOption("nickname")) {
            String nickname = call.getString("nickname");
            Crisp.setUserNickname(nickname);
        }
        if (call.hasOption("phone")) {
            String phone = call.getString("phone");
            Crisp.setUserPhone(phone);
        }
        if (call.hasOption("avatar")) {
            String avatar = call.getString("avatar");
            Crisp.setUserAvatar(avatar);
        }
        call.resolve();
    }

    @PluginMethod
    public void pushEvent(PluginCall call) {
        String name = call.getString("name");
        String color = call.getString("color");
        switch (color) {
            case "red":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.RED));
                break;
            case "orange":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.ORANGE));
                break;
            case "yellow":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.YELLOW));
                break;
            case "green":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.GREEN));
                break;
            case "purple":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.PURPLE));
                break;
            case "pink":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.PINK));
                break;
            case "brown":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.BROWN));
                break;
            case "grey":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.GREY));
                break;
            case "black":
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.BLACK));
                break;
            default:
                Crisp.pushSessionEvent(new SessionEvent(name, SessionEvent.Color.BLUE));
        }
        call.resolve();
    }

    @PluginMethod
    public void setCompany(PluginCall call) {
        String name = call.getString("name");
        String url = call.getString("url");
        String description = call.getString("description");
        JSArray employment = call.getArray("employment");
        JSArray geolocation = call.getArray("geolocation");
        try {
            Crisp.setUserCompany(
                new Company(
                    name,
                    new URL(url),
                    description,
                    new Employment(employment.get(0).toString(), employment.get(1).toString()),
                    new Geolocation(geolocation.get(0).toString(), geolocation.get(1).toString())
                )
            );
        } catch (JSONException e) {
            e.printStackTrace();
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        call.resolve();
    }

    @PluginMethod
    public void setInt(PluginCall call) {
        String key = call.getString("key");
        int value = call.getInt("value");
        Crisp.setSessionInt(key, value);
        call.resolve();
    }

    @PluginMethod
    public void setString(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");
        Crisp.setSessionString(key, value);
        call.resolve();
    }

    @PluginMethod
    public void sendMessage(PluginCall call) {
        call.unimplemented("Not implemented on Android.");
    }

    @PluginMethod
    public void setSegment(PluginCall call) {
        String segment = call.getString("segment");
        Crisp.setSessionSegment(segment);
        call.resolve();
    }

    @PluginMethod
    public void reset(PluginCall call) {
        Crisp.resetChatSession(this.getCrispContext());
        call.resolve();
    }

    @PluginMethod
    public void registerPushToken(PluginCall call) {
        String token = call.getString("token");
        if (token == null || token.isEmpty()) {
            call.reject("token is required");
            return;
        }
        CrispNotificationClient.sendTokenToCrisp(this.getCrispContext(), token);
        call.resolve();
    }

    @PluginMethod
    public void enableNotifications(PluginCall call) {
        Crisp.enableNotifications(this.getCrispContext(), true);
        call.resolve();
    }

    @PluginMethod
    public void isCrispPushNotification(PluginCall call) {
        Map<String, String> data = this.getNotificationData(call);
        JSObject ret = new JSObject();
        ret.put("isCrisp", CrispNotificationClient.isCrispNotification(data));
        call.resolve(ret);
    }

    @PluginMethod
    public void handlePushNotification(PluginCall call) {
        Map<String, String> data = this.getNotificationData(call);
        boolean openChatbox = call.getBoolean("openChatbox", true);
        boolean isCrisp = CrispNotificationClient.isCrispNotification(data);
        CrispNotificationClient.handleNotification(this.getCrispContext(), data, openChatbox);
        if (isCrisp && !openChatbox) {
            notifyCrispEvent("messageReceived", getPushMessageEvent());
        }
        call.resolve();
    }

    @PluginMethod
    public void setShouldPromptForNotificationPermission(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void openChatboxFromNotification(PluginCall call) {
        boolean opened = false;
        if (this.getActivity() != null) {
            opened = CrispNotificationClient.openChatbox(this.getActivity(), this.getActivity().getIntent());
        }
        JSObject ret = new JSObject();
        ret.put("opened", opened);
        call.resolve(ret);
    }

    private Map<String, String> getNotificationData(PluginCall call) {
        JSObject data = call.getObject("data");
        Map<String, String> map = new HashMap<>();
        if (data == null) {
            return map;
        }
        Iterator<String> keys = data.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            map.put(key, data.optString(key));
        }
        return map;
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}
